### PR TITLE
model: moca-embed/MoCa-Qwen25VL-3B (#3589)

### LINKED ISSUE #3589: Add model: moca-embed/MoCa-Qwen25VL-3B
### Model link on Hugging Face

https://huggingface.co/moca-embed/MoCa-Qwen25VL-3B

### Arxiv link

_No response_

### Description of the model

_No response_

### PR DESCRIPTION
**Pull request by hubielu**
Adds moca-embed/MoCa-Qwen25VL-3B (arXiv:2506.23115). Closes #3589.

**Implementation.** The reference repo vendors a fork of modeling_qwen2_5_vl.py, but the only functional changes are is_causal=False on the text attention and mean pooling instead of the LM head, so this wrapper does both on the stock Qwen2_5_VLModel. Pinned via a new moca extra to transformers>=4.49.0,<4.53 (same pattern as visrag-ret): v5 dropped _checkpoint_conversion_mapping and this checkpoint uses the pre-4.52 key layout, so it loads with silently missing weights, and _update_causal_mask is gone from 4.53 onward.

**Attention.** With flash_attention_2 the bidirectionality comes from module.is_causal; sdpa/eager still build a causal mask, so the wrapper patches _update_causal_mask to a padding-only mask and raises if that hook disappears. Verified on the model card example, all four values within 0.0020 (one bf16 ULP): 0.4824/0.4844, 0.3613/0.3594, 0.3945/0.3926, 0.3184/0.3203.

**Reproduction.** Vidore2ESGReportsHLRetrieval, english subset, ndcg@5 = 62.3 against 63.3 in Table 2 (ESG_Human). RTX A6000, sdpa, batch size 8.

**Prompt.** The image-side prompt is worth ~10 nDCG points here: model card wording gives 52.1, MoCa's own ViDoRe wording ("Represent the given document image.") gives 62.3. Per review that is now the default, so 62.3 needs no configuration; image_prompt remains a loader kwarg.

**Contamination.** MoCa's contrastive stage includes the ColPali training set, so ViDoRe-v1 is in-domain; ViDoRe-v2 was built after ColPali, hence used here. MMEB-train subsets with mteb equivalents are declared in training_datasets: HatefulMemes, SUN397, VOC2007, OK-VQA, CIRR, NIGHTS, WebQA, VisualNews, MSCOCO. N24News, A-OKVQA, Visual7W and VisDial have none. ChartQA, DocVQA, InfographicsVQA and ImageNet-1K exist only as differently-sourced packagings.

**Not covered.** 3B only (7B can follow), english subsets, all sdpa. Happy to run the other ViDoRe-v2 tasks or some of MMEB before merging.

### CODE REVIEW

File: mteb_mock_run_results.md
Comment (isaac-chung) [inline mteb_mock_run_results.md:1]: Can be removed.

File: mteb/models/model_implementations/moca_models.py [hunk 1]
```diff
@@ -0,0 +1,274 @@
...
+            all_embeddings.append(reps.cpu().to(torch.float32))
+
+        return torch.cat(all_embeddings, dim=0)
+
+
+moca_training_datasets = set(
+    # CPT: DCLM, PixelProse, MAmmoTH-VL-Instruct, DocMatix, VisRAG, ColPali train, MMEB train
+    # CL: VisRAG, ColPali train, MMEB train, mmE5, E5
+    # Not expressible as mteb task names, but note MMEB-train overlaps several MIEB
+    # tasks, so MIEB numbers for this model should be read as in-domain.
```
Comment (isaac-chung) [inline mteb/models/model_implementations/moca_models.py]: > MMEB-train overlaps several MIEB tasks, so MIEB numbers for this model should be read as in-domain.

Could you state which ones can be found in MTEB? 
https://huggingface.co/datasets/TIGER-Lab/MMEB-train

File: mteb/models/model_implementations/moca_models.py [hunk 2]
```diff
@@ -0,0 +1,274 @@
...
+                "the published numbers, which were produced with flash_attention_2.",
+                model_name,
+            )
+            if not hasattr(self.model.language_model, "_update_causal_mask"):
+                raise RuntimeError(
+                    "Cannot disable causal masking on transformers "
+                    f"{transformers.__version__} without flash attention: "
+                    "Qwen2_5_VLTextModel._update_causal_mask no longer exists. "
+                    "Install flash-attn, or pin transformers==4.52.4."
+                )
```
Comment (isaac-chung) [inline mteb/models/model_implementations/moca_models.py:144]: So either run with flash_attention_2 or run with older transformers. Any preference on how to handle this better here? @Samoed @KennethEnevoldsen 
It seems like the paper results need flash_attention_2

**Review by isaac-chung — COMMENTED** [review]
> The image-side prompt is worth about 10 nDCG points on this task. The model card wording ("Represent the given image.") gives 52.1, while the wording MoCa's own ViDoRe retriever uses ("Represent the given document image.") gives 62.3. I've exposed it as an image_prompt loader kwarg defaulting to the model card string, so document retrieval currently needs an explicit override. Would you prefer per-task prompt routing, or a different default? Happy to wire up whichever you'd like.

MoCa's own ViDoRe retriever as default ("Represent the given document image.") sounds more reasonable to me.

File: mteb/models/model_implementations/moca_models.py (hunk 2 above)
Comment (Samoed) [reply mteb/models/model_implementations/moca_models.py:144]: I think better to run with older transformers

File: mteb_mock_run_results.md
Comment (hubielu) [reply mteb_mock_run_results.md:1]: Removed

File: mteb/models/model_implementations/moca_models.py (hunk 2 above)
Comment (hubielu) [reply mteb/models/model_implementations/moca_models.py:144]: Sounds good, I went with older transformers. While doing it I found the pin was actually broken: extra_requirements_groups was ["transformers-v4"], which is transformers<5 and resolves to the newest 4.x — but _update_causal_mask is gone from 4.53 onward, so the RuntimeError in this wrapper would have fired for anyone installing it. So, I replaced it with a dedicated moca extra pinned >=4.49.0,<4.53, matching visrag-ret, and registered it in the same conflicts list. flash_attention_2 still works if installed and is still what the published numbers used.

File: mteb/models/model_implementations/moca_models.py (hunk 1 above)
Comment (hubielu) [reply mteb/models/model_implementations/moca_models.py]: Yes, I checked MMEB-train against the task registry. These have mteb equivalents and are now listed in training_datasets:

HatefulMemes → HatefulMemesI2TRetrieval, HatefulMemesT2IRetrieval
SUN397 → SUN397, SUN397ZeroShot
VOC2007 → VOC2007
OK-VQA → OKVQAIT2TRetrieval
CIRR → CIRRIT2IRetrieval
NIGHTS → NIGHTSI2IRetrieval
WebQA → WebQAT2ITRetrieval, WebQAT2TRetrieval
VisualNews → VisualNewsI2TRetrieval, VisualNewsT2IRetrieval
MSCOCO → MSCOCOI2TRetrieval, MSCOCOT2IRetrieval

No mteb equivalent: N24News, A-OKVQA, Visual7W, VisDial.

Left out deliberately: ChartQA, DocVQA and InfographicsVQA exist in mteb only as JinaVDR/ViDoRe/VisRAG packagings built from different sources, and ImageNet-1K only as ImageNet10Clustering / ImageNetDog15Clustering / TinyImageNetClustering.

**Comment by hubielu** [comment]
@isaac-chung Done — "Represent the given document image." is now the default, so the 62.3 in the description is what you get out of the box with no override. image_prompt stays as a loader kwarg for anyone who wants the model card wording.

**Comment by isaac-chung** [comment]
@hubielu thanks!

The last thing I believe is regenerating `uv.lock`, since `pyproject.toml` was updated. Otherwise this LGTM. 

[Bonus] If you still have the result files from the 3B runs on Vidore2ESGReportsHLRetrieval - it would also be great to add them to the results repo in a sep PR. Not required for this to merge.

**Comment by hubielu** [comment]
> @hubielu thanks!
> 
> The last thing I believe is regenerating `uv.lock`, since `pyproject.toml` was updated. Otherwise this LGTM.
> 
> [Bonus] If you still have the result files from the 3B runs on Vidore2ESGReportsHLRetrieval - it would also be great to add them to the results repo in a sep PR. Not required for this to merge.

Both done.

uv.lock regenerated. The diff is large but no packages changed, 0 added, 0 removed. It's all resolution-markers: adding a conflicting extra threads extra-4-mteb-moca through every marker string in the file.

Results are up in embeddings-benchmark/results#665.



Note that results#665 can't go green until this merges, its CI resolves model metadata from the installed mteb, so it fails with "not found in MTEB registry" until this one lands.

**Review by isaac-chung — APPROVED** [review]
Thanks!

**Comment by tomaarsen** [comment]
Hello!

I just saw this integration. I think this model would've also been very well suited for a Sentence Transformers integration on the model (repository) side. Then the MTEB integration would be simpler, and so would the usage for actual users. The integration would be similar as some of the models from https://huggingface.co/blog/multimodal-sentence-transformers#supported-multimodal-embedding-models.

- Tom Aarsen

**Comment by isaac-chung** [comment]
> Hello!
> 
> I just saw this integration. I think this model would've also been very well suited for a Sentence Transformers integration on the model (repository) side. Then the MTEB integration would be simpler, and so would the usage for actual users. The integration would be similar as some of the models from https://huggingface.co/blog/multimodal-sentence-transformers#supported-multimodal-embedding-models.
> 
> - Tom Aarsen 

That'd be very neat!
